cmd/k8s-deployment-scaler/main.go

```diff
@@ -0,0 +1,354 @@
+		writeInternalServerError(w, err)
+	}
+}
+
+// getDeploymentFromCache retrieves a deployment from the informer's cache
+func getDeploymentFromCache(namespace, name string) (*DeploymentInfo, bool) {
+	key := fmt.Sprintf("%s/%s", namespace, name)
+	obj, exists, err := deploymentInformer.GetIndexer().GetByKey(key)
+	if err != nil {
+		log.Printf("Error getting deployment from cache: %v", err)
+		return nil, false
+	}
+	if !exists {
+		return nil, false
+	}
+	deployment := obj.(*appsv1.Deployment)
+	return &DeploymentInfo{
+		Replicas:   *deployment.Spec.Replicas,
+		Deployment: deployment,
+	}, true
+}
+
+// handleGetReplicaCount handles the /replica-count endpoint for GET requests
+func handleGetReplicaCount(w http.ResponseWriter, r *http.Request) {
+	namespace := r.URL.Query().Get("namespace")
+	deploymentName := r.URL.Query().Get("deployment")
+
+	// Validate the query parameters
+	if namespace == "" || deploymentName == "" {
+		writeJSONError(w, apiError{
+			Message: "Both namespace and deployment must be specified",
+			Code:    http.StatusBadRequest,
+		})
+		return
+	}
+
+	// Get the deployment from the cache
+	info, exists := getDeploymentFromCache(namespace, deploymentName)
+	if !exists {
+		writeJSONError(w, apiError{
+			Message: "Deployment not found in cache",
+			Code:    http.StatusNotFound,
+		})
+		return
+	}
+
+	response := map[string]interface{}{
+		"replicaCount": info.Replicas,
+	}
+	if err := encodeAndWriteJSON(w, response); err != nil {
+		writeInternalServerError(w, err)
+	}
+}
+
+// handlePostReplicaCount handles the /replica-count endpoint for POST requests
+func handlePostReplicaCount(w http.ResponseWriter, r *http.Request) {
+	namespace := r.URL.Query().Get("namespace")
+	deploymentName := r.URL.Query().Get("deployment")
+
+	// Validate the query parameters
+	if namespace == "" || deploymentName == "" {
+		writeJSONError(w, apiError{
+			Message: "Missing query parameters",
+			Code:    http.StatusBadRequest,
+		})
+		return
+	}
+
+	var reqBody struct {
+		Replicas int32 `json:"replicas"`
+	}
+
+	// Decode the request body
+	if err := json.NewDecoder(r.Body).Decode(&reqBody); err != nil {
+		writeJSONError(w, apiError{
+			Message: "Invalid request body",
+			Code:    http.StatusBadRequest,
+		})
+		return
+	}
+
+	// Validate the replica count
+	if reqBody.Replicas < 0 {
+		writeJSONError(w, apiError{
+			Message: "Replica count must be non-negative",
+			Code:    http.StatusBadRequest,
+		})
+		return
+	}
+
+	// Get the current deployment from the cache
+	info, exists := getDeploymentFromCache(namespace, deploymentName)
+	if !exists {
+		writeJSONError(w, apiError{
+			Message: "Deployment not found in cache",
+			Code:    http.StatusNotFound,
+		})
+		return
+	}
+
+	// Update the deployment
+	info.Deployment.Spec.Replicas = &reqBody.Replicas
+	_, err := clientset.AppsV1().Deployments(namespace).Update(context.TODO(), info.Deployment, metav1.UpdateOptions{})
```

Comment:
Is there a better context you can use here?

Reply:
You're right. Context matters. I'll use the request's context with a timeout. It's cleaner and safer. Thanks for catching that.